File: dante/sockd/tests/logformat_iolog_test.py
#!/usr/bin/env python3
"""Check typed iolog events through the production formatter and serializer."""
import json
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest

BINARY = str(Path(sys.argv.pop(1)).resolve())


class IologTest(unittest.TestCase):
    def run_case(self, case, format="json"):
        with tempfile.TemporaryFile() as output:
            p = subprocess.run([BINARY, format, case], stdout=output,
                               stderr=subprocess.PIPE, timeout=15)
            self.assertEqual(p.returncode, 0, p.stderr)
            output.seek(0)
            return output.read()

    def events(self, case):
        return [json.loads(line) for line in self.run_case(case).splitlines()]

    def test_operations_and_rule(self):
        events = self.events("all")
        self.assertEqual([e["event"] for e in events], ["accept", "block", "temporary_block",
            "connect", "disconnect", "error", "temporary_error", "hostid", "io"])
        for e in events:
            self.assertEqual(e["rule"], {"number": 7, "type": "socks-rule"})
            self.assertEqual(e["protocol"], "tcp")
            self.assertEqual(e["command"], "connect")
            self.assertEqual(e["verdict"], "block" if e["event"] in ("block", "temporary_block") else "pass")
            self.assertNotIn("error", e)  # Explicit diagnostic is not proof of current errno.
            self.assertEqual(e["tcp_info"], 'first\nsecond "info"\\')
            if e["event"] != "io":
                self.assertEqual(e["detail"], 'detail "quoted"\nsecond line')
                self.assertNotIn("payload", e)

    def test_endpoints_proxies_and_authentication(self):
        e = self.events("all")[0]
        self.assertEqual(e["source"]["local"], dict(type="ipv4", address="127.0.0.1", port=1080))
        self.assertEqual(e["source"]["peer"], dict(type="ipv4", address="198.51.100.1", port=12345))
        self.assertEqual(e["source"]["authentication"], dict(method="username", user='alice"\\'))
        self.assertEqual(e["destination"]["local"], dict(type="ipv6", address="fe80::1", port=0, scope_id=3))
        self.assertEqual(e["destination"]["peer"], dict(type="domain", address="target.example", port=443))
        self.assertEqual(e["destination"]["authentication"], dict(method="none"))
        self.assertEqual(e["source_proxy"]["peer"]["address"], "192.0.2.1")
        self.assertEqual(e["destination_proxy"]["local"]["port"], 1082)
        self.assertNotIn("never-log-this-password", json.dumps(e))

    def test_unknown_and_zero(self):
        e, = self.events("missing")
        self.assertEqual(e["rule"]["number"], 0)
        for field in ("source", "destination", "source_proxy", "destination_proxy", "command", "detail"):
            self.assertNotIn(field, e)

    def test_ipv6_zero_scope_is_preserved(self):
        e = self.events("scopezero")[0]
        self.assertEqual(e["destination"]["local"]["scope_id"], 0)

    def test_known_error(self):
        import errno
        e, = self.events("error")
        self.assertEqual(e["error"], {"code": errno.ECONNREFUSED})
        self.assertEqual(e["event"], "error")
        self.assertEqual(e["verdict"], "pass")

    def test_payload_gating_and_zero(self):
        e, = self.events("payload")
        self.assertEqual(e["payload"], 'quote"\\\n\0ÿé')
        self.assertEqual(e["io_bytes"], 12)
        e, = self.events("ioonly")
        self.assertNotIn("payload", e)
        self.assertNotIn('quote', e["message"])
        self.assertEqual(e["io_bytes"], 12)
        e, = self.events("empty-payload")
        self.assertEqual(e["payload"], "")
        self.assertEqual(e["io_bytes"], 0)

    def test_filter_matrix_matches_raw(self):
        events = self.events("matrix")
        self.assertEqual(len(events), 18)
        raw = self.run_case("matrix", "raw").decode()
        # Remove exactly the stable raw prefix; message can itself be multiline.
        import re
        bodies = re.split(r"[^\n]*\(1789257600\.123456\) danted\[1234\]: info: ", raw)[1:]
        self.assertEqual([e["message"] for e in events], [s.removesuffix("\n") for s in bodies])
        self.assertTrue(all("tcp_info" not in e for e in events))

    def test_protocol_and_command(self):
        for case, protocol, command in (("udp", "udp", "udpassociate"), ("bind", "tcp", "bind")):
            for e in self.events(case):
                self.assertEqual((e["protocol"], e["command"]), (protocol, command))

    def test_long_payload_and_allocation_failure(self):
        for case in ("long", "alloc"):
            events = self.events(case)
            e = [e for e in events if e["event"] == "io"][0]
            self.assertTrue(e["truncated"])
            self.assertEqual(e["io_bytes"], 100000)
            self.assertLessEqual(len(self.run_case(case)), 65535 * 2)


if __name__ == "__main__":
    unittest.main()

File: dante/sockd/tests/logformat_iolog_test.c
/* Reuse the real logger and its deterministic OS boundary wrappers. */
#define main logger_driver_main
#include "logformat_logger_test.c"
#undef main

static void host(sockshost_t *value, int type, const char *address, int port)
{
   memset(value, 0, sizeof(*value));
   value->atype = type;
   value->port = htons(port);
   if (type == SOCKS_ADDR_IPV4)
      assert(inet_pton(AF_INET, address, &value->addr.ipv4) == 1);
   else if (type == SOCKS_ADDR_IPV6) {
      assert(inet_pton(AF_INET6, address, &value->addr.ipv6.ip) == 1);
      value->addr.ipv6.scopeid = 3;
   }
   else
      snprintf(value->addr.domain, sizeof(value->addr.domain), "%s", address);
}

int main(int argc, char **argv)
{
   rule_t rule;
   connectionstate_t state;
   iologaddr_t src, dst, sourceproxy, destproxy;
   const iologaddr_t *srcp = &src, *dstp = &dst;
   const iologaddr_t *spp = &sourceproxy, *dpp = &destproxy;
   const char payload[] = "quote\"\\\n\0\377\303\251";
   char huge[100001];
   const char *data = "detail \"quoted\"\nsecond line";
   size_t datalen = strlen(data);
   int out = STDOUT_FILENO, op, option, first = OPERATION_ACCEPT, last = OPERATION_IO;
   assert(argc == 3);
   __progname = "danted";
   sockscf.state.pid = 1234;
   sockscf.state.type = PROC_IO;
   sockscf.logformat = strcmp(argv[1], "json") == 0 ? LOGFORMAT_JSON : LOGFORMAT_RAW;
   sockscf.log.type = LOGTYPE_FILE;
   sockscf.log.filenoc = 1;
   sockscf.log.filenov = &out;
   memset(&rule, 0, sizeof(rule));
   memset(&state, 0, sizeof(state));
   memset(&src, 0, sizeof(src));
   memset(&dst, 0, sizeof(dst));
   memset(&sourceproxy, 0, sizeof(sourceproxy));
   memset(&destproxy, 0, sizeof(destproxy));
   rule.number = 7;
   rule.type = object_srule;
   rule.verdict = VERDICT_PASS;
   state.protocol = SOCKS_TCP;
   state.command = SOCKS_CONNECT;
   state.tcpinfo = "first\nsecond \"info\"\\";
   src.local_isset = src.peer_isset = src.auth_isset = 1;
   host(&src.local, SOCKS_ADDR_IPV4, "127.0.0.1", 1080);
   host(&src.peer, SOCKS_ADDR_IPV4, "198.51.100.1", 12345);
   src.auth.method = AUTHMETHOD_UNAME;
   snprintf((char *)src.auth.mdata.uname.name, sizeof(src.auth.mdata.uname.name), "alice\"\\");
   snprintf((char *)src.auth.mdata.uname.password, sizeof(src.auth.mdata.uname.password), "never-log-this-password");
   dst.local_isset = dst.peer_isset = dst.auth_isset = 1;
   host(&dst.local, SOCKS_ADDR_IPV6, "fe80::1", 0);
   host(&dst.peer, SOCKS_ADDR_DOMAIN, "target.example", 443);
   dst.auth.method = AUTHMETHOD_NONE;
   sourceproxy.peer_isset = 1;
   host(&sourceproxy.peer, SOCKS_ADDR_IPV4, "192.0.2.1", 1081);
   destproxy.local_isset = 1;
   host(&destproxy.local, SOCKS_ADDR_IPV4, "192.0.2.2", 1082);
#if HAVE_SOCKS_HOSTID
   src.hostidc = 1;
   assert(inet_pton(AF_INET, "203.0.113.1", &src.hostidv[0]) == 1);
#endif
   if (strcmp(argv[2], "missing") == 0) {
      memset(&src, 0, sizeof(src));
      dstp = spp = dpp = NULL;
      rule.number = 0;
      state.command = SOCKS_UNKNOWN;
      data = NULL;
      datalen = 0;
      first = last = OPERATION_ACCEPT;
   }
   else if (strcmp(argv[2], "error") == 0) {
      data = NULL;
      datalen = 0;
      first = last = OPERATION_ERROR;
   }
   else if (strcmp(argv[2], "payload") == 0 || strcmp(argv[2], "ioonly") == 0) {
      data = payload;
      datalen = sizeof(payload) - 1;
      first = last = OPERATION_IO;
   }
   else if (strcmp(argv[2], "empty-payload") == 0) {
      data = "";
      datalen = 0;
      first = last = OPERATION_IO;
   }
   else if (strcmp(argv[2], "long") == 0 || strcmp(argv[2], "alloc") == 0) {
      memset(huge, 'x', sizeof(huge) - 1);
      huge[sizeof(huge) - 1] = '\0';
      data = huge;
      datalen = sizeof(huge) - 1;
      first = last = OPERATION_IO;
      if (strcmp(argv[2], "alloc") == 0)
         failalloc = 1;
   }
   else if (strcmp(argv[2], "scopezero") == 0)
      dst.local.addr.ipv6.scopeid = 0;
   else if (strcmp(argv[2], "udp") == 0) {
      state.protocol = SOCKS_UDP;
      state.command = SOCKS_UDPASSOCIATE;
   }
   else if (strcmp(argv[2], "bind") == 0)
      state.command = SOCKS_BIND;

   for (option = 0; option < (strcmp(argv[2], "matrix") == 0 ? 7 : 1); ++option) {
      memset(&rule.log, 0, sizeof(rule.log));
      if (strcmp(argv[2], "matrix") == 0) {
         rule.log.connect = option == 1;
         rule.log.disconnect = option == 2;
         rule.log.error = option == 3;
         rule.log.data = option == 4;
         rule.log.iooperation = option == 5;
         rule.log.tcpinfo = option == 6;
      }
      else {
         rule.log.connect = rule.log.disconnect = rule.log.error = 1;
         rule.log.data = strcmp(argv[2], "ioonly") != 0;
         rule.log.iooperation = 1;
         rule.log.tcpinfo = 1;
      }
      for (op = first; op <= last; ++op) {
         errno = ECONNREFUSED;
         iolog(&rule, &state, op, srcp, dstp, spp, dpp, data, datalen);
         if (sockscf.logformat == LOGFORMAT_JSON)
            assert(errno == ECONNREFUSED);
      }
   }
   return 0;
}
